### PR TITLE
Miscellaneous options cleanup

### PR DESCRIPTION
**Pull request by jasonmalinowski**
Some boring changes that are hopefully quick and mechanical to review. Reviewing commit by commit might be recommended:

1. We had some duplicate and unused options due a mismerge. Let's delete those.
2. We had some options that forwarded to other options. Let delete those.
3. Document.Options, a property I added earlier, really should be asynchronous since that requires walking files. Make it so. There's one potential perf concern since we're waiting for it on the UI thread, but the intent is in practice for open files it'll already be available and the function will complete synchronously. The only case where it should actually be async is either we just opened the file in VS and might still be processing configuration files, or the file is closed.
4. Simplify the parts of OptionSet you can overload. Although this may look like a breaking change, it's not from our last released version.

*Review:* @dotnet/roslyn-ide. This PR is designed to be a quick review, so something fast would be appreciated.

### CODE REVIEW

File: src/EditorFeatures/CSharp/AutomaticCompletion/AutomaticLineEnderCommandHandler.cs [hunk 1]
```diff
@@ -97,7 +97,9 @@ protected override void FormatAndApply(Document document, int position, Cancella
                 return;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharp/AutomaticCompletion/AutomaticLineEnderCommandHandler.cs]: what the heck is the AutomaticLineEnderCommandHandler?!

File: src/EditorFeatures/CSharp/AutomaticCompletion/Sessions/CurlyBraceCompletionSession.cs [hunk 2]
```diff
@@ -143,7 +143,19 @@ private void FormatTrackingSpan(IBraceCompletionSession session, bool shouldHono
                 }
             }
 
-            if (session.SubjectBuffer.GetOption(SmartIndent) == IndentStyle.Smart)
+            var document = snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            IndentStyle style;
+
+            if (document != null)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharp/AutomaticCompletion/Sessions/CurlyBraceCompletionSession.cs]: ternary expression?

File: src/Workspaces/Core/Portable/PublicAPI.Shipped.txt [hunk 3]
```diff
@@ -515,11 +515,11 @@ Microsoft.CodeAnalysis.Options.OptionKey.Option.get -> Microsoft.CodeAnalysis.Op
 Microsoft.CodeAnalysis.Options.OptionKey.OptionKey(Microsoft.CodeAnalysis.Options.IOption option, string language = null) -> void
 Microsoft.CodeAnalysis.Options.OptionSet
 abstract Microsoft.CodeAnalysis.Options.OptionSet.GetOption(Microsoft.CodeAnalysis.Options.OptionKey optionKey) -> object
-abstract Microsoft.CodeAnalysis.Options.OptionSet.GetOption<T>(Microsoft.CodeAnalysis.Options.Option<T> option) -> T
-abstract Microsoft.CodeAnalysis.Options.OptionSet.GetOption<T>(Microsoft.CodeAnalysis.Options.PerLanguageOption<T> option, string language) -> T
+Microsoft.CodeAnalysis.Options.OptionSet.GetOption<T>(Microsoft.CodeAnalysis.Options.Option<T> option) -> T
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/PublicAPI.Shipped.txt]: do these need to be virtual to not be a breaking change?

**Comment by CyrusNajmabadi** [comment]
I'm a bit concerned about the blocking on the UI thread.  But overall this all looks good.

File: src/EditorFeatures/CSharpTest/Formatting/Indentation/SmartIndenterEnterOnTokenTests.cs [hunk 4]
```diff
@@ -1365,7 +1365,7 @@ void Main(object o)
                 Assert.True(
                     CSharpIndentationService.ShouldUseSmartTokenFormatterInsteadOfIndenter(
                         Formatter.GetDefaultFormattingRules(workspace, root.Language),
-                        root, line.AsTextLine(), document.Options, CancellationToken.None));
+                        root, line.AsTextLine(), await document.GetOptionsAsync(CancellationToken.None), CancellationToken.None));
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Formatting/Indentation/SmartIndenterEnterOnTokenTests.cs]: presumably GetOptionsAsync takes an optional cancellation token.  If so, feel free to pass nothing in tests.

File: src/EditorFeatures/Core/Implementation/AutomaticCompletion/AbstractAutomaticLineEnderCommandHandler.cs [hunk 5]
```diff
@@ -78,7 +78,7 @@ public void ExecuteCommand(AutomaticLineEnderCommandArgs args, Action nextHandle
             }
 
             // feature off
-            if (!document.Options.GetOption(InternalFeatureOnOffOptions.AutomaticLineEnder))
+            if (!document.Project.Solution.Workspace.Options.GetOption(InternalFeatureOnOffOptions.AutomaticLineEnder))
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/AutomaticCompletion/AbstractAutomaticLineEnderCommandHandler.cs]: is this not using GetOptionsAsync because it's an internal option?

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Presentation/CompletionPresenter.cs [hunk 6]
```diff
@@ -45,7 +45,7 @@ internal sealed class CompletionPresenter : ForegroundThreadAffinitizedObject, I
 
             var document = subjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
 
-            var completionSetFactory = document != null && NeedsDev15CompletionSetFactory(document.Options, document.Project.Language)
+            var completionSetFactory = document != null && NeedsDev15CompletionSetFactory(document.Project.Solution.Options, document.Project.Language)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Presentation/CompletionPresenter.cs]: how can Solution directly expose options?  wouldn't is be affected by .editorconfig files above the .sln file?

File: src/Features/Core/Portable/CodeFixes/NamingStyle/AbstractNamingStyleCodeFixProvider.cs [hunk 7]
```diff
@@ -46,7 +46,7 @@ public override ImmutableArray<string> FixableDiagnosticIds
                             solution,
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/CodeFixes/NamingStyle/AbstractNamingStyleCodeFixProvider.cs]: can you extract the outer lambda into it's own function.

File: src/Features/Core/Portable/CodeRefactorings/ExtractMethod/AbstractExtractMethodCodeRefactoringProvider.cs [hunk 8]
```diff
@@ -66,7 +66,8 @@ internal class ExtractMethodCodeRefactoringProvider : CodeRefactoringProvider
 
             if (result.Succeeded || result.SucceededWithSuggestion)
             {
-                var description = document.Options.GetOption(ExtractMethodOptions.AllowMovingDeclaration) ?
+                var documentOptions = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+                var description = documentOptions.GetOption(ExtractMethodOptions.AllowMovingDeclaration) ?
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/CodeRefactorings/ExtractMethod/AbstractExtractMethodCodeRefactoringProvider.cs]: put ? and : on the next lines.

File: src/Features/Core/Portable/Shared/Extensions/DocumentExtensions.cs [hunk 9]
```diff
@@ -14,7 +14,8 @@ internal static class DocumentExtensions
     {
         public static bool ShouldHideAdvancedMembers(this Document document)
         {
-            return document.Options.GetOption(CompletionOptions.HideAdvancedMembers);
+            // Since we don't actually have a way to configure this per-document, we can fetch from the core workspace
+            return document.Project.Solution.Workspace.Options.GetOption(CompletionOptions.HideAdvancedMembers, document.Project.Language);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Shared/Extensions/DocumentExtensions.cs]: what's the intuition for what is a document option vs Workspace option?

File: src/Features/Core/Portable/SpellCheck/AbstractSpellCheckCodeFixProvider.cs [hunk 10]
```diff
@@ -57,8 +57,8 @@ internal abstract class AbstractSpellCheckCodeFixProvider<TSimpleName> : CodeFix
             // Disable snippets from ever appearing in the completion items. It's
             // very unlikely the user would ever mispell a snippet, then use spell-
             // checking to fix it, then try to invoke the snippet.
-            var options = document.Options.WithChangedOption(
-                CompletionOptions.SnippetsBehavior, document.Project.Language, SnippetsRule.NeverInclude);
+            var options = (await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false))
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/SpellCheck/AbstractSpellCheckCodeFixProvider.cs]: in product code i really don't like awaits in subexpressions like this.  It's really hard to parse.  I'd prefer to just break into multiple statements.

File: src/Workspaces/Core/Portable/Workspace/Solution/Document.cs [hunk 11]
```diff
@@ -460,13 +460,12 @@ private string GetDebuggerDisplay()
         /// Returns the options that should be applied to this document. This consists of global options from <see cref="Solution.Options"/>,
         /// merged with any settings the user has specified at the solution, project, and document levels.
         /// </summary>
-        public DocumentOptionSet Options
+        /// <remarks>
+        /// This method is async because this may require reading other files. It is expected this is cheap and synchronous in most cases due to caching.</remarks>
+        public Task<DocumentOptionSet> GetOptionsAsync(CancellationToken cancellationToken)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/Document.cs]: CancellationToken should be optional here since this is public.

File: src/Workspaces/Core/Portable/PublicAPI.Shipped.txt (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/PublicAPI.Shipped.txt]: i read your comment now.  You're fine hten :)

File: src/Features/Core/Portable/Shared/Extensions/DocumentExtensions.cs (hunk 9 above)
Comment (dpoeschl) [inline src/Features/Core/Portable/Shared/Extensions/DocumentExtensions.cs]: My basic understanding was that editorconfig was for things impacting the actual code, not the ide experience (e.g. this option that just effects the completion list experience). This matches with the list of [generally-adopted editorconfig properties](https://github.com/editorconfig/editorconfig/wiki/EditorConfig-Properties). But maybe Jason had other reasons.

File: src/EditorFeatures/CSharp/AutomaticCompletion/AutomaticLineEnderCommandHandler.cs (hunk 1 above)
Comment (jasonmalinowski) [inline src/EditorFeatures/CSharp/AutomaticCompletion/AutomaticLineEnderCommandHandler.cs]: It's the shift+enter thing where you escape all your braces or something.

File: src/Features/Core/Portable/Shared/Extensions/DocumentExtensions.cs (hunk 9 above)
Comment (jasonmalinowski) [inline src/Features/Core/Portable/Shared/Extensions/DocumentExtensions.cs]: @dpoeschl's understanding is correct @CyrusNajmabadi. When I originally did the split, I just moved everything I could to document options because "it was easy". Now that document options are marginally harder to get and potentially involves blocking, I've generally gone back to a model where on/off flags for features are just fetched from Workspace options, since we don't plan to expose editorconfig values for those.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Presentation/CompletionPresenter.cs (hunk 6 above)
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Presentation/CompletionPresenter.cs]: I guess this was leftover. Let me just delete it.

File: src/EditorFeatures/Core/Implementation/AutomaticCompletion/AbstractAutomaticLineEnderCommandHandler.cs (hunk 5 above)
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/AutomaticCompletion/AbstractAutomaticLineEnderCommandHandler.cs]: It's not document specific. See other comment from @dpoeschl.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Presentation/CompletionPresenter.cs (hunk 6 above)
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Presentation/CompletionPresenter.cs]: Actually I'm going to leave this: looks like we have a big of a silly bug historically where options could change while we're asynchronously doing stuff, and places that requested options that way would be bad. This will let us reestablish snapshot semantics. It doesn't do that yet, but once I had a WorkspaceChangedKind for options changed, we'll fix this up.

Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Presentation/CompletionPresenter.cs]: I'll update the comment though.
